cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/DirectRuntimeRequestValidator.java

```diff
@@ -73,48 +73,49 @@
     long pollTimeMillis = cConf.getLong(Constants.RuntimeMonitor.POLL_TIME_MS);
     this.programRunsCache = CacheBuilder.newBuilder()
       .expireAfterWrite(pollTimeMillis, TimeUnit.MILLISECONDS)
-      .build(new CacheLoader<ProgramRunId, Boolean>() {
+      .build(new CacheLoader<ProgramRunId, RunRecordDetail>() {
         @Override
-        public Boolean load(ProgramRunId programRunId) throws IOException, UnauthorizedException {
-          return isValid(programRunId);
+        public RunRecordDetail load(ProgramRunId programRunId) throws UnauthorizedException, NotFoundException,
+          IOException {
+          return runRecord(programRunId);
         }
       });
   }
 
   @Override
-  public void validate(ProgramRunId programRunId, HttpRequest request) throws BadRequestException {
+  public RunRecordDetail checkRunRecord(ProgramRunId programRunId, HttpRequest request) throws BadRequestException {
     accessEnforcer.enforce(programRunId, authenticationContext.getPrincipal(), StandardPermission.GET);
-    boolean exists;
+    RunRecordDetail runRecord;
     try {
-      exists = programRunsCache.get(programRunId);
+      runRecord = programRunsCache.get(programRunId);
+      return runRecord;
     } catch (Exception e) {
       throw new ServiceUnavailableException(Constants.Service.RUNTIME, e);
```

Comment:
Why it is service unavailable?

Reply:
Actually, I am keeping the current behavior intact. The existing code also throws a ServiceUnavailbleException - https://github.com/cdapio/cdap/blob/develop/cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/DirectRuntimeRequestValidator.java#L90 - so there is no change in the method contract with new changes